### PR TITLE
Dynamic parameter setting of costmap layers (inflation/static/voxel)

### PR DESCRIPTION
**Pull request by indraneelpatil**
---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | ros-planning#956 |
| Primary OS tested on | Ubuntu |
| Robotic platform tested on | Turtlebot3 simulation |

---

## Description of contribution in a few bullet points

* Dynamic parameter setting callbacks are added in the layers using suitable mutexes to prevent segfaults
* Decision on whether parameters are settable/non-settable dynamically are taken after referring to these layers in the ROS1 nav stack and warning messages are added wherever parameter setting is rejected
* Appropriate functions are called after dynamic parameter changes (also referring to the ROS1 nav stack)
* Changes are tested on the turtlebot3 simulation
* Suitable test cases are added to test parameter setting

## Description of documentation updates required from your changes

---

#### For Maintainers: 
- [ ] Check that any new parameters added are updated in navigation.ros.org
- [ ] Check that any significant change is added to the migration guide
- [ ] Check that any new features **OR** changes to existing behaviors are reflected in the tuning guide
- [ ] Check that any new functions have Doxygen added
- [ ] Check that any new features have test coverage
- [ ] Check that any new plugins is added to the plugins page
- [ ] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists

### CODE REVIEW

**Review by AlexeyMerzlyakov — COMMENTED** [review]
@indraneelpatil, thank you for your work on covering parameters and making it able to be changed on-the-fly!
Some general thoughts. This patch is made for `InflationLayer`, `StaticLayer` and `VoxelLayer`. `ObstacleLayer` and `RangeSensorLayer` are out of scope. So, they could be covered by separate PR. However, since `VoxelLayer` is inherited from `ObstacleLayer` class, they have much in common. So, the main question: should `ObstacleLayer` better be handled by this PR as well (also see some comments from this review), or leave it for next PR?

Also, on this stage, I've made the review for the code. Testcases and coverage will be checked at next step, when the review items will be discussed and met.

File: nav2_costmap_2d/plugins/inflation_layer.cpp [hunk 1]
```diff
@@ -407,4 +415,77 @@ InflationLayer::generateIntegerDistances()
...
+
+  if (paramChange_s.enabled || paramChange_s.inflate_unknown ||
+    paramChange_s.inflate_around_unknown || paramChange_s.inflation_radius ||
+    paramChange_s.cost_scaling_factor)
+  {
+    need_reinflation_ = true;
+  }
+
+  if (paramChange_s.inflation_radius || paramChange_s.cost_scaling_factor) {
+    computeCaches();
```
[the remark this answers is not in the document]
Comment (indraneelpatil) [reply nav2_costmap_2d/plugins/inflation_layer.cpp:484]: Hey! both computeCaches() and dynamicParametersCallback are using the getMutex() function which is returning a recursive_mutex called access_

So I think the deadlock situation is already taken care of?

File: nav2_costmap_2d/plugins/inflation_layer.cpp [hunk 2]
```diff
@@ -407,4 +415,77 @@ InflationLayer::generateIntegerDistances()
...
+  std::lock_guard<Costmap2D::mutex_t> guard(*getMutex());
+
+  struct paramChange
+  {
+    bool inflation_radius : 1;
+    bool cost_scaling_factor : 1;
+    bool enabled : 1;
+    bool inflate_unknown : 1;
+    bool inflate_around_unknown : 1;
+  } paramChange_s{0, 0, 0, 0, 0};
```
[the remark this answers is not in the document]
Comment (indraneelpatil) [reply nav2_costmap_2d/plugins/inflation_layer.cpp:436]: Done

File: nav2_costmap_2d/plugins/inflation_layer.cpp [hunk 3]
```diff
@@ -407,4 +415,77 @@ InflationLayer::generateIntegerDistances()
...
+    }
+
+  }
+
+  if (paramChange_s.enabled || paramChange_s.inflate_unknown ||
+    paramChange_s.inflate_around_unknown || paramChange_s.inflation_radius ||
+    paramChange_s.cost_scaling_factor)
+  {
+    need_reinflation_ = true;
+  }
```
[the remark this answers is not in the document]
Comment (indraneelpatil) [reply nav2_costmap_2d/plugins/inflation_layer.cpp:481]: Done

File: nav2_costmap_2d/plugins/inflation_layer.cpp [hunk 4]
```diff
@@ -407,4 +415,77 @@ InflationLayer::generateIntegerDistances()
...
+
+      } else if (param_name == name_ + "." + "cost_scaling_factor" &&
+        cost_scaling_factor_ != parameter.as_double())
+      {
+        cost_scaling_factor_ = parameter.as_double();
+        paramChange_s.cost_scaling_factor = true;
+      }
+    } else if (param_type == ParameterType::PARAMETER_BOOL) {
+      if (param_name == name_ + "." + "enabled" && enabled_ != parameter.as_bool()) {
+        enabled_ = parameter.as_bool();
```
[the remark this answers is not in the document]
Comment (indraneelpatil) [reply nav2_costmap_2d/plugins/inflation_layer.cpp:459]: Done

File: nav2_costmap_2d/plugins/inflation_layer.cpp [hunk 5]
```diff
@@ -407,4 +415,77 @@ InflationLayer::generateIntegerDistances()
...
+    const auto & param_name = parameter.get_name();
+
+
+    if (param_type == ParameterType::PARAMETER_DOUBLE) {
+      if (param_name == name_ + "." + "inflation_radius" &&
+        inflation_radius_ != parameter.as_double())
+      {
+        inflation_radius_ = parameter.as_double();
+        paramChange_s.inflation_radius = true;
+        cell_inflation_radius_ = cellDistance(inflation_radius_);
```
[the remark this answers is not in the document]
Comment (indraneelpatil) [reply nav2_costmap_2d/plugins/inflation_layer.cpp:449]: Done

File: nav2_costmap_2d/plugins/static_layer.cpp [hunk 6]
```diff
@@ -435,4 +444,46 @@ StaticLayer::updateCosts(
...
+        transform_tolerance_ = tf2::durationFromSec(parameter.as_double());
+      }
+    } else if (param_type == ParameterType::PARAMETER_BOOL) {
+      if (param_name == name_ + "." + "enabled" && enabled_ != parameter.as_bool()) {
+        enabled_ = parameter.as_bool();
+
+        x_ = y_ = 0;
+        width_ = size_x_;
+        height_ = size_y_;
+        has_updated_data_ = true;
```
[the remark this answers is not in the document]
Comment (indraneelpatil) [reply nav2_costmap_2d/plugins/static_layer.cpp:480]: Okay done

File: nav2_costmap_2d/include/nav2_costmap_2d/static_layer.hpp [hunk 7]
```diff
@@ -178,6 +185,9 @@ class StaticLayer : public CostmapLayer
   tf2::Duration transform_tolerance_;
   std::atomic<bool> update_in_progress_;
   nav_msgs::msg::OccupancyGrid::SharedPtr map_buffer_;
+  // Dynamic parameters handler
+  rclcpp::node_interfaces::OnSetParametersCallbackHandle::SharedPtr dyn_params_handler_;
+  std::mutex dyn_param_mutex_;
```
[the remark this answers is not in the document]
Comment (indraneelpatil) [reply nav2_costmap_2d/include/nav2_costmap_2d/static_layer.hpp:190]: Done

File: nav2_costmap_2d/plugins/voxel_layer.cpp [hunk 8]
```diff
@@ -458,4 +468,60 @@ void VoxelLayer::updateOrigin(double new_origin_x, double new_origin_y)
...
+      } else if (param_name == name_ + "." + "mark_threshold") {
+        mark_threshold_ = parameter.as_int();
+      } else if (param_name == name_ + "." + "combination_method") {
+        combination_method_ = parameter.as_int();
+      }
+    }
+
+  }
+
+  matchSize();
```
[the remark this answers is not in the document]
Comment (indraneelpatil) [reply nav2_costmap_2d/plugins/voxel_layer.cpp:522]: Done

File: nav2_costmap_2d/include/nav2_costmap_2d/voxel_layer.hpp [hunk 9]
```diff
@@ -220,6 +221,18 @@ class VoxelLayer : public ObstacleLayer
...
+
+  /**
+   * @brief Callback executed when a parameter change is detected
+   * @param event ParameterEvent message
+   */
+  rcl_interfaces::msg::SetParametersResult
+  dynamicParametersCallback(std::vector<rclcpp::Parameter> parameters);
+
+  // Dynamic parameters handler
+  std::mutex mutex_;
```
[the remark this answers is not in the document]
Comment (indraneelpatil) [reply nav2_costmap_2d/include/nav2_costmap_2d/voxel_layer.hpp:233]: Done

File: nav2_costmap_2d/plugins/voxel_layer.cpp [hunk 10]
```diff
@@ -458,4 +468,60 @@ void VoxelLayer::updateOrigin(double new_origin_x, double new_origin_y)
...
+      } else if (param_name == name_ + "." + "publish_voxel_map") {
+        RCLCPP_WARN(
+          logger_, "publish voxel map is not a dynamic parameter "
+          "cannot be changed while running. Rejecting parameter update.");
+        continue;
+      }
+
+    } else if (param_type == ParameterType::PARAMETER_INTEGER) {
+      if (param_name == name_ + "." + "z_voxels") {
+        size_z_ = parameter.as_int();
```
[the remark this answers is not in the document]
Comment (indraneelpatil) [reply nav2_costmap_2d/plugins/voxel_layer.cpp:510]: Done

File: nav2_costmap_2d/plugins/voxel_layer.cpp [hunk 11]
```diff
@@ -458,4 +468,60 @@ void VoxelLayer::updateOrigin(double new_origin_x, double new_origin_y)
...
+      } else if (param_name == name_ + "." + "unknown_threshold") {
+        unknown_threshold_ = parameter.as_int() + (VOXEL_BITS - size_z_);
+      } else if (param_name == name_ + "." + "mark_threshold") {
+        mark_threshold_ = parameter.as_int();
+      } else if (param_name == name_ + "." + "combination_method") {
+        combination_method_ = parameter.as_int();
```
[the remark this answers is not in the document]
Comment (indraneelpatil) [reply nav2_costmap_2d/plugins/voxel_layer.cpp:516]: Done

File: nav2_costmap_2d/plugins/voxel_layer.cpp [hunk 12]
```diff
@@ -458,4 +468,60 @@ void VoxelLayer::updateOrigin(double new_origin_x, double new_origin_y)
...
+      } else if (param_name == name_ + "." + "origin_z") {
+        origin_z_ = parameter.as_double();
+      } else if (param_name == name_ + "." + "z_resolution") {
+        z_resolution_ = parameter.as_double();
+      }
+    } else if (param_type == ParameterType::PARAMETER_BOOL) {
+      if (param_name == name_ + "." + "enabled") {
+        enabled_ = parameter.as_bool();
+      } else if (param_name == name_ + "." + "footprint_clearing_enabled") {
+        footprint_clearing_enabled_ = parameter.as_bool();
```
[the remark this answers is not in the document]
Comment (indraneelpatil) [reply nav2_costmap_2d/plugins/voxel_layer.cpp:500]: Done

File: nav2_costmap_2d/plugins/voxel_layer.cpp [hunk 13]
```diff
@@ -458,4 +468,60 @@ void VoxelLayer::updateOrigin(double new_origin_x, double new_origin_y)
...
+      if (param_name == name_ + "." + "max_obstacle_height") {
+        max_obstacle_height_ = parameter.as_double();
+      } else if (param_name == name_ + "." + "origin_z") {
+        origin_z_ = parameter.as_double();
+      } else if (param_name == name_ + "." + "z_resolution") {
+        z_resolution_ = parameter.as_double();
+      }
+    } else if (param_type == ParameterType::PARAMETER_BOOL) {
+      if (param_name == name_ + "." + "enabled") {
+        enabled_ = parameter.as_bool();
```
[the remark this answers is not in the document]
Comment (indraneelpatil) [reply nav2_costmap_2d/plugins/voxel_layer.cpp:497]: Done

File: nav2_costmap_2d/plugins/voxel_layer.cpp [hunk 14]
```diff
@@ -458,4 +468,60 @@ void VoxelLayer::updateOrigin(double new_origin_x, double new_origin_y)
...
+  std::lock_guard<std::mutex> lock_reinit(mutex_);
+
+  for (auto parameter : parameters) {
+    const auto & param_type = parameter.get_type();
+    const auto & param_name = parameter.get_name();
+
+    if (param_type == ParameterType::PARAMETER_DOUBLE) {
+
+      if (param_name == name_ + "." + "max_obstacle_height") {
+        max_obstacle_height_ = parameter.as_double();
```
[the remark this answers is not in the document]
Comment (indraneelpatil) [reply nav2_costmap_2d/plugins/voxel_layer.cpp:489]: List of parameters which are from the parent ObstacleLayer class

- combination_method_
- footprint_clearing_enabled_
- max_obstacle_height_

In the latest commit we have added the mutex locks to the parent ObstacleLayer class so these parameters now seem well protected, so I think we can keep them. Do you insist on their removal?

File: nav2_costmap_2d/plugins/inflation_layer.cpp [hunk 15]
```diff
@@ -108,6 +115,7 @@ InflationLayer::onInitialize()
   need_reinflation_ = false;
   cell_inflation_radius_ = cellDistance(inflation_radius_);
   matchSize();
+
```
[the remark this answers is not in the document]
Comment (indraneelpatil) [reply nav2_costmap_2d/plugins/inflation_layer.cpp:118]: Done

File: nav2_costmap_2d/plugins/inflation_layer.cpp [hunk 16]
```diff
@@ -407,4 +415,77 @@ InflationLayer::generateIntegerDistances()
...
+    bool enabled : 1;
+    bool inflate_unknown : 1;
+    bool inflate_around_unknown : 1;
+  } paramChange_s{0, 0, 0, 0, 0};
+
+  for (auto parameter : parameters) {
+    const auto & param_type = parameter.get_type();
+    const auto & param_name = parameter.get_name();
+
+
```
[the remark this answers is not in the document]
Comment (indraneelpatil) [reply nav2_costmap_2d/plugins/inflation_layer.cpp:438]: Done

**Comment by indraneelpatil** [comment]
@AlexeyMerzlyakov Thanks for the great review, I have addressed all comments in the latest commits and ensured test cases are still passing, would love to hear back from your whenever possible

File: nav2_costmap_2d/plugins/inflation_layer.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (AlexeyMerzlyakov) [reply nav2_costmap_2d/plugins/inflation_layer.cpp:473]: Yes, it is already covered by that needed. That was my oversight.

File: nav2_costmap_2d/test/integration/inflation_tests.cpp [hunk 17]
```diff
@@ -600,3 +601,48 @@ TEST_F(TestNode, testInflation3)
...
   ASSERT_EQ(countValues(*costmap, nav2_costmap_2d::INSCRIBED_INFLATED_OBSTACLE), 4u);
 }
+
+
+/**
+ * Test dynamic parameter setting of inflation layer
+ */
+TEST_F(TestNode, testDynParamsSet)
+{
+  auto node = std::make_shared<rclcpp_lifecycle::LifecycleNode>("inflation_dyn_param_test");
```
Comment (AlexeyMerzlyakov) [inline nav2_costmap_2d/test/integration/inflation_tests.cpp:611]: This parameter is unused. Why do we need to keep it?

File: nav2_costmap_2d/test/integration/obstacle_tests.cpp [hunk 18]
```diff
@@ -405,5 +406,103 @@ TEST_F(TestNode, testRaytracing) {
...
 
   ASSERT_EQ(lethal_count, 1);
+}
+
+/**
+ * Test dynamic parameter setting of voxel layer
+ */
+TEST_F(TestNode, testDynParamsSetVoxel)
+{
+  auto node = std::make_shared<rclcpp_lifecycle::LifecycleNode>("voxel_dyn_param_test");
```
Comment (AlexeyMerzlyakov) [inline nav2_costmap_2d/test/integration/obstacle_tests.cpp:416]: Again, this looks like an unnecessary code.

File: nav2_costmap_2d/test/integration/obstacle_tests.cpp [hunk 19]
```diff
@@ -405,5 +406,103 @@ TEST_F(TestNode, testRaytracing) {
...
+    rclcpp::Parameter("voxel_layer.combination_method", 0),
+    rclcpp::Parameter("voxel_layer.mark_threshold", 1),
+    rclcpp::Parameter("voxel_layer.unknown_threshold", 10),
+    rclcpp::Parameter("voxel_layer.z_resolution", 0.4),
+    rclcpp::Parameter("voxel_layer.origin_z", 1.0),
+    rclcpp::Parameter("voxel_layer.z_voxels", 14),
+    rclcpp::Parameter("voxel_layer.max_obstacle_height", 4.0),
+    rclcpp::Parameter("voxel_layer.footprint_clearing_enabled", false),
+    rclcpp::Parameter("voxel_layer.enabled", false)
+  });
```
Comment (AlexeyMerzlyakov) [inline nav2_costmap_2d/test/integration/obstacle_tests.cpp:449]: One corner case is not being covered:
Screenshot 2022-01-28 at 16-57-20 LCOV - dynamic_param_1 info - plugins voxel_layer cpp

File: nav2_costmap_2d/test/integration/obstacle_tests.cpp [hunk 20]
```diff
@@ -405,5 +406,103 @@ TEST_F(TestNode, testRaytracing) {
...
+
+}
+
+
+/**
+ * Test dynamic parameter setting of static layer
+ */
+TEST_F(TestNode, testDynParamsSetStatic)
+{
+  auto node = std::make_shared<rclcpp_lifecycle::LifecycleNode>("static_dyn_param_test");
```
Comment (AlexeyMerzlyakov) [inline nav2_costmap_2d/test/integration/obstacle_tests.cpp:477]: The same comment (extra-code).

File: nav2_costmap_2d/test/integration/obstacle_tests.cpp [hunk 21]
```diff
@@ -405,5 +406,103 @@ TEST_F(TestNode, testRaytracing) {
...
+  auto parameter_client = std::make_shared<rclcpp::AsyncParametersClient>(
+    costmap->get_node_base_interface(), costmap->get_node_topics_interface(),
+    costmap->get_node_graph_interface(),
+    costmap->get_node_services_interface());
+
+  auto results = parameter_client->set_parameters_atomically(
+  {
+    rclcpp::Parameter("static_layer.transform_tolerance", 1.0),
+    rclcpp::Parameter("static_layer.enabled", false)
+  });
```
Comment (AlexeyMerzlyakov) [inline nav2_costmap_2d/test/integration/obstacle_tests.cpp:495]: Some corner cases are not being covered (as follows):
Screenshot 2022-01-28 at 16-56-32 LCOV - dynamic_param_1 info - plugins static_layer cpp

**Review by AlexeyMerzlyakov — COMMENTED** [review]
Sorry for my late reply due to some sick last week.
Regarding the change, this seems is much better. Some few strokes before we go.

Regarding separation of `ObstacleLayer`, I personally would like to concatenate these two into one PR, but also don't mind for your option. The change for `ObstacleLayer` could take a new Observation API, so this should be paid an attention in the next step as separate PR. I am personally OK abut it too.

File: nav2_costmap_2d/plugins/inflation_layer.cpp [hunk 22]
```diff
@@ -407,4 +417,65 @@ InflationLayer::generateIntegerDistances()
...
+  * @param event ParameterEvent message
+  */
+rcl_interfaces::msg::SetParametersResult
+InflationLayer::dynamicParametersCallback(
+  std::vector<rclcpp::Parameter> parameters)
+{
+  static bool need_cache_recompute = false;
+
+  rcl_interfaces::msg::SetParametersResult result;
+  std::lock_guard<Costmap2D::mutex_t> guard(*getMutex());
```
[the remark this answers is not in the document]
Comment (indraneelpatil) [reply nav2_costmap_2d/plugins/inflation_layer.cpp:431]: Sounds good!

File: nav2_costmap_2d/plugins/voxel_layer.cpp [hunk 23]
```diff
@@ -458,4 +468,68 @@ void VoxelLayer::updateOrigin(double new_origin_x, double new_origin_y)
...
+  * @brief Callback executed when a parameter change is detected
+  * @param event ParameterEvent message
+  */
+rcl_interfaces::msg::SetParametersResult
+VoxelLayer::dynamicParametersCallback(
+  std::vector<rclcpp::Parameter> parameters)
+{
+  static bool resize_map_needed = false;
+  rcl_interfaces::msg::SetParametersResult result;
+  std::lock_guard<Costmap2D::mutex_t> guard(*getMutex());
```
[the remark this answers is not in the document]
Comment (indraneelpatil) [reply nav2_costmap_2d/plugins/voxel_layer.cpp:481]: Done

File: nav2_costmap_2d/plugins/inflation_layer.cpp (hunk 22 above)
[the remark this answers is not in the document]
Comment (indraneelpatil) [reply nav2_costmap_2d/plugins/inflation_layer.cpp:431]: Done

File: nav2_costmap_2d/plugins/static_layer.cpp [hunk 24]
```diff
@@ -435,4 +444,47 @@ StaticLayer::updateCosts(
...
+/**
+  * @brief Callback executed when a parameter change is detected
+  * @param event ParameterEvent message
+  */
+rcl_interfaces::msg::SetParametersResult
+StaticLayer::dynamicParametersCallback(
+  std::vector<rclcpp::Parameter> parameters)
+{
+  rcl_interfaces::msg::SetParametersResult result;
+  std::lock_guard<Costmap2D::mutex_t> guard(*getMutex());
```
[the remark this answers is not in the document]
Comment (indraneelpatil) [reply nav2_costmap_2d/plugins/static_layer.cpp:456]: Done

File: nav2_costmap_2d/include/nav2_costmap_2d/voxel_layer.hpp [hunk 25]
```diff
@@ -51,6 +51,7 @@
 #include <message_filters/subscriber.h>
 #include <nav2_costmap_2d/obstacle_layer.hpp>
 #include <nav2_voxel_grid/voxel_grid.hpp>
+#include <mutex>
```
[the remark this answers is not in the document]
Comment (indraneelpatil) [reply nav2_costmap_2d/include/nav2_costmap_2d/voxel_layer.hpp:54]: Done

File: nav2_costmap_2d/test/integration/inflation_tests.cpp (hunk 17 above)
Comment (indraneelpatil) [reply nav2_costmap_2d/test/integration/inflation_tests.cpp:611]: Aaah, thanks great spot!

File: nav2_costmap_2d/test/integration/obstacle_tests.cpp (hunk 18 above)
Comment (indraneelpatil) [reply nav2_costmap_2d/test/integration/obstacle_tests.cpp:416]: Done!

File: nav2_costmap_2d/test/integration/obstacle_tests.cpp (hunk 20 above)
Comment (indraneelpatil) [reply nav2_costmap_2d/test/integration/obstacle_tests.cpp:477]: Done!

File: nav2_costmap_2d/test/integration/obstacle_tests.cpp (hunk 19 above)
Comment (indraneelpatil) [reply nav2_costmap_2d/test/integration/obstacle_tests.cpp:449]: Done

File: nav2_costmap_2d/test/integration/obstacle_tests.cpp (hunk 21 above)
Comment (indraneelpatil) [reply nav2_costmap_2d/test/integration/obstacle_tests.cpp:495]: Done!

**Comment by indraneelpatil** [comment]
@AlexeyMerzlyakov I covered the latest review! Hoping to hear back from you soon!

File: nav2_costmap_2d/test/integration/obstacle_tests.cpp [hunk 26]
```diff
@@ -405,5 +406,107 @@ TEST_F(TestNode, testRaytracing) {
...
+  EXPECT_EQ(costmap->get_parameter("voxel_layer.z_voxels").as_int(), 14);
+  EXPECT_EQ(costmap->get_parameter("voxel_layer.max_obstacle_height").as_double(), 4.0);
+  EXPECT_EQ(costmap->get_parameter("voxel_layer.footprint_clearing_enabled").as_bool(), false);
+  EXPECT_EQ(costmap->get_parameter("voxel_layer.enabled").as_bool(), false);
+  EXPECT_EQ(costmap->get_parameter("voxel_layer.publish_voxel_map").as_bool(), true);
+
+  costmap->on_deactivate(rclcpp_lifecycle::State());
+  costmap->on_cleanup(rclcpp_lifecycle::State());
+  costmap->on_shutdown(rclcpp_lifecycle::State());
+
```
Comment (AlexeyMerzlyakov) [inline nav2_costmap_2d/test/integration/obstacle_tests.cpp:468]: Extra blank

File: nav2_costmap_2d/test/integration/inflation_tests.cpp [hunk 27]
```diff
@@ -600,3 +601,46 @@ TEST_F(TestNode, testInflation3)
...
+  EXPECT_EQ(costmap->get_parameter("inflation_layer.inflation_radius").as_double(), 0.0);
+  EXPECT_EQ(costmap->get_parameter("inflation_layer.cost_scaling_factor").as_double(), 0.0);
+  EXPECT_EQ(costmap->get_parameter("inflation_layer.inflate_unknown").as_bool(), true);
+  EXPECT_EQ(costmap->get_parameter("inflation_layer.inflate_around_unknown").as_bool(), true);
+  EXPECT_EQ(costmap->get_parameter("inflation_layer.enabled").as_bool(), false);
+
+  costmap->on_deactivate(rclcpp_lifecycle::State());
+  costmap->on_cleanup(rclcpp_lifecycle::State());
+  costmap->on_shutdown(rclcpp_lifecycle::State());
+
```
Comment (AlexeyMerzlyakov) [inline nav2_costmap_2d/test/integration/inflation_tests.cpp:645]: Extra blank

**Review by AlexeyMerzlyakov — APPROVED** [review]
All previous items were met. Please remove extra blank lines (pointed out where) and ready for merge.

File: nav2_costmap_2d/plugins/static_layer.cpp [hunk 28]
```diff
@@ -435,4 +444,48 @@ StaticLayer::updateCosts(
...
 
+/**
+  * @brief Callback executed when a parameter change is detected
+  * @param event ParameterEvent message
+  */
+rcl_interfaces::msg::SetParametersResult
+StaticLayer::dynamicParametersCallback(
+  std::vector<rclcpp::Parameter> parameters)
+{
+
```
[the remark this answers is not in the document]
Comment (indraneelpatil) [reply nav2_costmap_2d/plugins/static_layer.cpp:455]: Done

File: nav2_costmap_2d/plugins/inflation_layer.cpp [hunk 29]
```diff
@@ -407,4 +417,65 @@ InflationLayer::generateIntegerDistances()
...
 
+/**
+  * @brief Callback executed when a parameter change is detected
+  * @param event ParameterEvent message
+  */
+rcl_interfaces::msg::SetParametersResult
+InflationLayer::dynamicParametersCallback(
+  std::vector<rclcpp::Parameter> parameters)
+{
+
```
[the remark this answers is not in the document]
Comment (indraneelpatil) [reply nav2_costmap_2d/plugins/inflation_layer.cpp:428]: Done

File: nav2_costmap_2d/test/integration/obstacle_tests.cpp [hunk 30]
```diff
@@ -405,5 +406,107 @@ TEST_F(TestNode, testRaytracing) {
...
+}
+
+
```
[the remark this answers is not in the document]
Comment (indraneelpatil) [reply nav2_costmap_2d/test/integration/obstacle_tests.cpp:471]: Done

File: nav2_costmap_2d/test/integration/obstacle_tests.cpp (hunk 26 above)
Comment (indraneelpatil) [reply nav2_costmap_2d/test/integration/obstacle_tests.cpp:468]: Done

File: nav2_costmap_2d/test/integration/obstacle_tests.cpp [hunk 31]
```diff
@@ -405,5 +406,107 @@ TEST_F(TestNode, testRaytracing) {
...
+  EXPECT_EQ(costmap->get_parameter("static_layer.transform_tolerance").as_double(), 1.0);
+  EXPECT_EQ(costmap->get_parameter("static_layer.enabled").as_bool(), false);
+  EXPECT_EQ(costmap->get_parameter("static_layer.map_subscribe_transient_local").as_bool(), false);
+  EXPECT_EQ(costmap->get_parameter("static_layer.map_topic").as_string(), "dynamic_topic");
+  EXPECT_EQ(costmap->get_parameter("static_layer.subscribe_to_updates").as_bool(), true);
+
+  costmap->on_deactivate(rclcpp_lifecycle::State());
+  costmap->on_cleanup(rclcpp_lifecycle::State());
+  costmap->on_shutdown(rclcpp_lifecycle::State());
```
[the remark this answers is not in the document]
Comment (indraneelpatil) [reply nav2_costmap_2d/test/integration/obstacle_tests.cpp:511]: Done

File: nav2_costmap_2d/test/integration/inflation_tests.cpp [hunk 32]
```diff
@@ -600,3 +601,46 @@ TEST_F(TestNode, testInflation3)
   ASSERT_EQ(countValues(*costmap, nav2_costmap_2d::LETHAL_OBSTACLE), 1u);
   ASSERT_EQ(countValues(*costmap, nav2_costmap_2d::INSCRIBED_INFLATED_OBSTACLE), 4u);
 }
+
+
```
[the remark this answers is not in the document]
Comment (indraneelpatil) [reply nav2_costmap_2d/test/integration/inflation_tests.cpp:605]: Done

File: nav2_costmap_2d/test/integration/inflation_tests.cpp (hunk 27 above)
Comment (indraneelpatil) [reply nav2_costmap_2d/test/integration/inflation_tests.cpp:645]: Done

**Comment by indraneelpatil** [comment]
@AlexeyMerzlyakov Removed blank lines, thanks!

**Comment by AlexeyMerzlyakov** [comment]
@SteveMacenski, I think, we could merge this.

File: nav2_costmap_2d/plugins/inflation_layer.cpp [hunk 33]
```diff
@@ -113,6 +120,7 @@ InflationLayer::onInitialize()
 void
 InflationLayer::matchSize()
 {
+  std::lock_guard<Costmap2D::mutex_t> guard(*getMutex());
```
Comment (SteveMacenski) [inline nav2_costmap_2d/plugins/inflation_layer.cpp:123]: My understanding of the mutex in the inflation layer is that it is for locking in the inflation resource (see the doxygen comment on it) and the previous implementation was only locking the update / values in the inflation cost data structure. It feels like locking it also for the parameters is abusing the role here a little.

 Would it make more sense to have these variables as atomics instead?

File: nav2_costmap_2d/plugins/obstacle_layer.cpp [hunk 34]
```diff
@@ -455,6 +455,7 @@ ObstacleLayer::updateFootprint(
   double * max_x,
   double * max_y)
 {
+  std::lock_guard<Costmap2D::mutex_t> guard(*getMutex());
```
Comment (SteveMacenski) [inline nav2_costmap_2d/plugins/obstacle_layer.cpp:458]: Why is this costmap layer not having any dynamic parameters? I see the voxel does, which derives from this, and would implicitly add those dynamic parameters to both sets of layers at once

Comment (indraneelpatil) [reply nav2_costmap_2d/plugins/obstacle_layer.cpp:458]: Yes, so we originally started off the PR with Obstacle Layer out of scope of these changes (ref Alexey's comments above) but then we realised that it wasnt possible to make voxel dynamic without also making obstacle layer dynamic.

But then we decided that since this PR is already pretty big, reviewing it and keeping it in sync with master is becoming a problem so we think its best if we target ObstacleLayer in the next PR. Is this okay?

**Comment by indraneelpatil** [comment]
@SteveMacenski  Yes we used the turtlebot3 simulation to try all these parameter changes out!

File: nav2_costmap_2d/plugins/obstacle_layer.cpp (hunk 34 above)
Comment (SteveMacenski) [reply nav2_costmap_2d/plugins/obstacle_layer.cpp:458]: OK!

File: nav2_costmap_2d/plugins/inflation_layer.cpp (hunk 33 above)
Comment (SteveMacenski) [reply nav2_costmap_2d/plugins/inflation_layer.cpp:123]: Good point. Maybe its better to use another mutex then -- or do you think its OK that we're kind of abusing this mutex? Its not really for the same thing anymore. At bare minimum, the description of the mutex needs updating.

Comment (AlexeyMerzlyakov) [reply nav2_costmap_2d/plugins/inflation_layer.cpp:123]: For InflationLayer, standard mutex was chosen to cover `inflation_radius_` and `need_reinflation_` variables shared usage by `matchSize()`, `updateBounds()`, `onFootprintChanged()` and `dynamicParametersCallback()`. However, introducing of new lock for them will cause to have 2 mutexes in `updateCosts()` + `computeCaches()` or `dynamicParametersCallback()` which is also accessing `cell_inflation_radius_` parameter used by already covered by standard mutex routines. Therefore, I've suggested to use one mutex for everything for simplicity.

Comment (SteveMacenski) [reply nav2_costmap_2d/plugins/inflation_layer.cpp:123]: OK, I think that's fine

File: nav2_costmap_2d/plugins/inflation_layer.cpp [hunk 35]
```diff
@@ -407,4 +417,64 @@ InflationLayer::generateIntegerDistances()
...
+        inflate_around_unknown_ != parameter.as_bool())
+      {
+        inflate_around_unknown_ = parameter.as_bool();
+        need_reinflation_ = true;
+      }
+    }
+
+  }
+
+  if (need_cache_recompute) {
```
Comment (SteveMacenski) [inline nav2_costmap_2d/plugins/inflation_layer.cpp:472]: Wouldn't it be better at this point to call `matchSize` which will handle `computeCaches()` and do all the `cellDistance()` stuff for you? That way the if/else conditions above can be more simple.

**Review by SteveMacenski — APPROVED** [review]

File: nav2_costmap_2d/plugins/inflation_layer.cpp (hunk 35 above)
Comment (indraneelpatil) [reply nav2_costmap_2d/plugins/inflation_layer.cpp:472]: I guess you are right, the decision of computeCache() or matchSize() I just made based on the ROS1 Nav stack https://github.com/ros-planning/navigation/blob/noetic-devel/costmap_2d/plugins/inflation_layer.cpp

Practically I think the only difference between the two approaches is that through matchSize() we are additionally resetting the resolution_ and seen vectors. But it doesnt seem like that causes any issues. Pushing the change.

**Review by SteveMacenski — APPROVED** [review]
@AlexeyMerzlyakov merge if you're happy with it, I am happy with it. You took the lead on it so I'll let you hit the button :-) 

@indraneelpatil thanks for the help and the repeated review cycles! I appreciate this feature so much and we're crazy close to being done with all this dynamic parameter infrastructure! I definitely look foward to the obstacle layer's.

**Comment by AlexeyMerzlyakov** [comment]
Also, looks good to me. @indraneelpatil, thank you for your well-made work! Looking forward for next part of dynamic parameters PR.
